src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/util/AbstractIncrementalClusterReducer.java

```diff
@@ -0,0 +1,87 @@
+package uk.ac.ebi.pride.spectracluster.hadoop.util;
+
+import uk.ac.ebi.pride.spectracluster.cluster.ICluster;
+import uk.ac.ebi.pride.spectracluster.engine.IIncrementalClusteringEngine;
+import uk.ac.ebi.pride.spectracluster.engine.IncrementalClusteringEngineFactory;
+import uk.ac.ebi.pride.spectracluster.spectrum.ISpectrum;
+import uk.ac.ebi.pride.spectracluster.util.ClusterUtilities;
+import uk.ac.ebi.pride.spectracluster.util.Defaults;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * AbstractIncrementalClusterReducer clusters using incremental clustering engine.
+ *
+ * @author Rui Wang
+ * @version $Id$
+ */
+public abstract class AbstractIncrementalClusterReducer extends FilterSingleSpectrumClusterReducer {
+    private final IncrementalClusteringEngineFactory engineFactory = new IncrementalClusteringEngineFactory();
+    private IIncrementalClusteringEngine engine;
+    private double clusterRetainThreshold = Defaults.getRetainThreshold();
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException {
+        super.setup(context);
+
+        // read and customize configuration, default will be used if not provided
+        ConfigurableProperties.configureAnalysisParameters(context.getConfiguration());
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException {
+        updateEngine(context, null);
+        super.cleanup(context);
+    }
+
+    protected abstract <T> void updateEngine(final Context context, final T key) throws IOException, InterruptedException;
+
+
+    /**
+     * Before writing out a cluster, remove all the non-fitting spectra and each out as a single-spectrum cluster
+     */
+    protected void writeCluster(Context context, ICluster cluster) throws IOException, InterruptedException {
```

Comment:
We could think about trying to cluster these non-fitting spectra again. Maybe they form a cluster. I'm not sure myself whether that's a sensible idea though.